servicetalk-encoding-netty/src/main/java/io/servicetalk/encoding/netty/NettyChannelContentCodec.java

```diff
@@ -0,0 +1,335 @@
+                            subscriber.onNext(buffer);
+                        }
+                    }
+
+                    @Override
+                    public void onError(Throwable t) {
+                        safeCleanup(channel);
+                        subscriber.onError(t);
+                    }
+
+                    @Override
+                    public void onComplete() {
+                        cleanup(channel);
+                        subscriber.onComplete();
+                    }
+                });
+    }
+
+    @Override
+    public Buffer decode(final Buffer src, final int offset, final int length, final BufferAllocator allocator) {
+        requireNonNull(src);
+        final int availableBytes = src.readableBytes() - offset;
+        if (offset < 0) {
+            throw new IllegalArgumentException("Invalid offset: " + offset + " (expected >= 0)");
+        }
+
+        if (length <= 0) {
+            throw new IllegalArgumentException("Invalid length: " + length + " (expected > 0)");
+        }
+
+        if (length > availableBytes) {
+            throw new IllegalStateException("Invalid length: " + length + " (expected <= " + availableBytes);
+        }
+
+        final ByteToMessageDecoder decoder = decoderSupplier.get();
+        final EmbeddedChannel channel = newEmbeddedChannel(decoder, allocator);
+
+        try {
+            ByteBuf origin = toByteBuf(src, offset, length);
+            channel.writeInbound(origin);
+            return newBufferFrom(channel.readInbound());
+        } catch (Throwable e) {
+            LOGGER.error("Error while decoding with {}", name(), e);
+            throwException(e);
+            return EMPTY_BUFFER;
+        } finally {
+            safeCleanup(channel);
+        }
+    }
+
+    @Override
+    public Publisher<Buffer> decode(final Publisher<Buffer> from, final BufferAllocator allocator) {
+        requireNonNull(from);
+        return from.liftSync(subscriber -> new PublisherSource.Subscriber<Buffer>() {
+
+            private final ByteToMessageDecoder decoder = decoderSupplier.get();
+            private final EmbeddedChannel channel = newEmbeddedChannel(decoder, allocator);
+
+            @Nullable
+            PublisherSource.Subscription subscription;
+
+            @Override
+            public void onSubscribe(final PublisherSource.Subscription subscription) {
+                this.subscription = subscription;
+                subscriber.onSubscribe(subscription);
+            }
+
+            @Override
+            public void onNext(@Nullable final Buffer src) {
+                assert subscription != null;
+
+                if (src == null) {
+                    return;
+                }
+
+                // onNext will produce AT-MOST N items (as received)
+                if (!channel.isOpen()) {
+                    throw new IllegalStateException("Stream encoder previously closed but more input arrived ");
```

Comment:
Consider moving this check to the top. We should not request more items if the channel is already closed

Reply:
Moved it further up, but it was already guarding further demand.